kernel/src/main/java/org/kframework/parser/concrete2kore/kernel/Scanner.java

```diff
@@ -39,10 +39,18 @@
 
     private static final String FLEX_LIB = OS.current().equals(OS.OSX) ? "-ll" : "-lfl";
 
+    private String layout;
+
     public Scanner(ParseInModule module) {
-        tokens = KSyntax2GrammarStatesFilter.getTokens(module.getParsingModule());
-        scanner = getScanner();
-        this.module = module.seedModule();
+        this.tokens  = KSyntax2GrammarStatesFilter.getTokens(module.getParsingModule());
+        this.module  = module.seedModule();
+        String whites = "[\\ \\n\\r\\t]";
+        if (this.module.layout().length() == 0) {
+            this.layout = whites;
```

Comment:
Can you help me understand what went wrong without this?

Reply:
Well, the symptom of the problem was that with no layout present, it would throw `Array access out of bounds` errors on line 285 for the tests in `ParserTest.java` (which does not go through `RuleGrammarGenerator.java`, and so does not get default layout added).

Originally it was throwing Flex errors, because the layout line being added to Flex was `() ;\n`, which it did not like. So I added the check that the layout was empty to `Scanner`, and did not put the line in unless it was non-empty. This fixed the Flex errors that were being thrown, but led into the array out of bounds access errors. As a test, to see if it was because it was expecting that line to be there in Flex, I changed the behaviour of layout to *always* include the whitespace at least, so that we would have something to put. It seems to have worked, fixing the tests in `ParserTest` while allowing custom comment layout (but throwing away whitespace).

One more thing to note: I peeked at the generated Flex files, and noticed that now that we have sort `#Layout` in the actual grammar, it's adding rules for those regexs to produce tokens in sort `#Layout`. This might give us a different approach to the problem, where we let the Scanner produce the tokens of sort `#Layout` for us, and we do a post-processing pass to throw them away. This might make it easier to eventually handle white-space sensitive languages like Python in the future.

Reply:
Unfortunately it's simply not feasible for us to always treat whitespace as layout, even in a non-indent-sensitive language. A language like visual basic uses newline to delineate statements, but doesn't use indentation to delineate blocks, which means that this issue is the only blocking issue preventing us from defining parsers for such languages in K. Given that this is clearly some kind of underlying bug, I'd like to try to figure it out.

Reply:
I should clarify, line 285 or `Scanner.java` was the one throwing the error, on the tests in `ParserTest`.